src/database-postgresql.cpp

```diff
@@ -0,0 +1,505 @@
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	const char* param_values[2];
+	int param_format[2];
+	int param_length[2];
+
+	std::string old_data = loadBlock(pos);
+
+	// This is of interest. I have noticed there are a large number of
+	// saveBlock calls where the new and old data are exactly the same.
+	if (data.compare(old_data) == 0) {
+		if (old_data.size() > 0) {
+			infostream << "saveBlock called for block already saved, pos = "
+					<< PP(pos) << ", " << data.size() << " bytes"
+					<< std::endl;
+		}
+		return true;
+	}
+
+	if (old_data.size() > 0) {
+		return updateBlock(pos, data);
+	}
+
+	deleteBlock(pos);
+
+	param_values[0] = posTochar(pos);
+	param_format[0] = 0;
+	param_length[0] = strlen(param_values[0]);
+
+	param_values[1] = data.c_str();
+	param_format[1] = 1;
+	param_length[1] = data.size();
+
+	m_res = PQexecPrepared(m_conn, "write_block", 2, (const char *const*)param_values,
+			param_length, param_format, 1); 
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		PQclear(m_res);
+		return true;
+	default:
+		errorstream << "PostgreSQL Error Message:" << PQresultErrorMessage(m_res) << std::endl;
+		PQclear(m_res);
+		return false;
+		break;
+	}
+}
+
+bool Database_PostgreSQL::updateBlock(const v3s16 &pos, const std::string &data)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	const char* param_values[2];
+	int param_format[2];
+	int param_length[2];
+
+	param_values[0] = data.c_str();
+	param_format[0] = 1;
+	param_length[0] = data.size();
+
+	param_values[1] = posTochar(pos);
+	param_format[1] = 0;
+	param_length[1] = strlen(param_values[0]);
+
+	m_res = PQexecPrepared(m_conn, "update_block", 2, (const char *const*)param_values,
+			param_length, param_format, 1); 
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		PQclear(m_res);
+		return true;
+	default:
+		errorstream << "PostgreSQL Error Message:" << PQresultErrorMessage(m_res) << std::endl;
+		PQclear(m_res);
+		return false;
+		break;
+	}
+}
+
+std::string Database_PostgreSQL::loadBlock(const v3s16 &pos)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	const char* param_values[1];
+	int param_format[1];
+	int param_length[1];
+
+	std::string s   = "";
+
+	param_values[0] = posTochar(pos);
+	param_format[0] = 0;
+	param_length[0] = strlen(param_values[0]);
+
+	m_res = PQexecPrepared(m_conn, "read_block", 1, (const char *const*)param_values, param_length, param_format, 1); 
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		break;
+	case PGRES_TUPLES_OK:
+		if (PQntuples(m_res))
+			s = std::string(PQgetvalue(m_res, 0,0), PQgetlength(m_res, 0, 0));
+		break;
+	default:
+		errorstream << "PostgreSQL Error:" << PQresultErrorMessage(m_res) << std::endl;
+		break;
+	}
+
+	PQclear(m_res);
+	return s;
+}
+
+const char * Database_PostgreSQL::posTochar(const v3s16 &pos)
+{
+	char buffer[33];
+	snprintf(buffer, 32, "%ld", getBlockAsInteger(pos));
```

Comment:
BTW, found another HUGE mistake here, I wonder how it even worked on a server. One must not return the pointer to a **stack array**. You should read about stack and heap objects. Th buffer has to be properly allocated and deleted via malloc/free or new and delete[].

Use `writeS16()` from `serialize.h`, make it like:
```
char *Database_PostgreSQL::posTochar(const v3s16 &pos)
{
char *result = new char[6];
writeS16((u8 *)result, pos.X);
writeS16((u8 *)result+2, pos.Y);
writeS16((u8 *)result+4, pos.Z);
return result;
}
```
then use it like normal, just after you passed the values to the postgresql call, do e.g.
`delete [] param_values[0];`.
OK that *would* work if param_values weren't an array of **const** char* (which, im not sure about which implicit conversions are possible, has to stay that way). Therefore you should do it like:
```
char *char_pos = posTochar(...);
param_values[0] = char_pos;
call_to_postgresql(..., param_values, ...);
delete [] char_pos;
```